core/src/main/scala/org/apache/spark/FutureAction.scala

```diff
@@ -154,17 +143,13 @@ class SimpleFutureAction[T] private[spark](jobWaiter: JobWaiter[_], resultFunc:
   override def isCancelled: Boolean = _cancelled
 
   override def value: Option[Try[T]] = {
-    if (jobWaiter.jobFinished) {
-      Some(awaitResult())
-    } else {
+    if (!isCompleted) {
       None
-    }
-  }
-
-  private def awaitResult(): Try[T] = {
-    jobWaiter.awaitResult() match {
-      case JobSucceeded => scala.util.Success(resultFunc)
-      case JobFailed(e: Exception) => scala.util.Failure(e)
+    } else {
+      jobWaiter.awaitResult() match {
```

Comment:
This part seems like a bad hack to use `awaitResult` to get the result. Rather, there should be a `JobWaiter.jobResult` (make it public), that return `Option[JobResult]` and use that.